### PR TITLE
Make open/close HDK callable

### PR DESCRIPTION
**Pull request by ThetaSinner**
### TODO:
- [x] CHANGELOGs updated with appropriate info

### CODE REVIEW

File: crates/test_utils/wasm/wasm_workspace/migrate_new/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+[package]
+name = "test_wasm_migrate_new"
+description = "Test wasm for DNA migration, intended to be the migration target"
+version = "0.0.1"
+authors = [ "thedavidmeister", "thedavidmeister@gmail.com" ]
```
Comment (jost-s) [inline crates/test_utils/wasm/wasm_workspace/migrate_new/Cargo.toml:5]: He keeps writing stuff even in his absence!

**Review by jost-s — APPROVED** [review]
Very nice! I wonder if this will work for an app with two DNAs as well. Seems like it should.

File: crates/test_utils/wasm/wasm_workspace/migrate_new/Cargo.toml (hunk 1 above)
Comment (ThetaSinner) [reply crates/test_utils/wasm/wasm_workspace/migrate_new/Cargo.toml]: Haha, the dangers of copy paste to get the features right for these crates - tidied up :)

**Comment by ThetaSinner** [comment]
> I wonder if this will work for an app with two DNAs as well.

Yeah I don't see why it wouldn't. I think migration has quite a long way to go though, like this is relying on knowing exactly what DNA is being migrated from which the conductor could inform the new DNA of... we'll see how that shapes up!

File: crates/holochain/src/core/sys_validate/error.rs [hunk 2]
```diff
@@ -210,6 +210,8 @@ pub enum PrevActionErrorKind {
     InvalidRoot,
     #[error("Root of source chain must have a timestamp greater than the Dna's origin_time")]
     InvalidRootOriginTime,
+    #[error("No more actions are allowed after a chain close")]
+    ActionAfterChainClose,
```
Comment (maackle) [inline crates/holochain/src/core/sys_validate/error.rs:214]: Should there be a parallel validation that says a ChainOpen should not be allowed after any non-genesis action?

File: crates/holochain/tests/migration.rs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+//! for migrating from one app version to another.
+
+use holo_hash::ActionHash;
+use holochain::sweettest::{SweetAgents, SweetConductor, SweetDnaFile};
+use holochain_serialized_bytes::prelude::*;
+use holochain_wasm_test_utils::TestWasm;
+use holochain_zome_types::prelude::{random_network_seed, YamlProperties};
+
+#[tokio::test(flavor = "multi_thread")]
+async fn migrate_dna_with_second_app_install() {
```
Comment (maackle) [inline crates/holochain/tests/migration.rs:12]: this is exciting to see! :tada:

File: crates/test_utils/wasm/wasm_workspace/migrate_new/src/coordinator.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        .unwrap()
+        .modifiers
+        .properties
+        .try_into()
+        .map_err(|_| {
+            wasm_error!(WasmErrorInner::Guest(
+                "Could not deserialize properties".to_string()
+            ))
+        })?;
+    open_chain(properties.prev_dna_hash.clone())?;
```
Comment (maackle) [inline crates/test_utils/wasm/wasm_workspace/migrate_new/src/coordinator.rs:21]: This is something we talked about the other day. We can't expect the prev DNA hash to exist in the DNA properties in real apps, so how is this supposed to come in? Seems we'd have to do the open_chain in a zome call which can receive that data from the client?

Comment (ThetaSinner) [reply crates/test_utils/wasm/wasm_workspace/migrate_new/src/coordinator.rs:21]: I think we can expect it to be present if we require a DNA to declare its lineage. So not part of the properties but somewhere on the DNA manifest. I think working out which DNA in the lineage, that the new version supports migration from, was actually present should be the conductors job and available somehow to the init callback.

File: crates/holochain/src/core/sys_validate/error.rs (hunk 2 above)
Comment (ThetaSinner) [reply crates/holochain/src/core/sys_validate/error.rs:214]: I'm not sure we should restrict this until we work out how it's going to be used. If it's used as part of init and done first then we could check that. But if we decide we want to support migrations being done after init, through some other zome function call, then this wouldn't necessarily appear right after genesis

**Pull request merged**